### PR TITLE
Adding API to support external app authenticate with Power Platform Connector Plugins

### PR DESCRIPTION
**Pull request by ndangudubiyyam**
For more information about how to contribute to this repo, visit [this page](https://github.com/OfficeDev/microsoft-teams-library-js/blob/main/CONTRIBUTING.md).

## Description

Adds 1P internal-only externalAppAuthentication.authenticateWithPowerPlatformConnectorPlugins api for delegating auth to the host. It signals to the host to perform PPC authentication for the application specified by the titleId and pluginId

> If this Pull Request should close/resolve any issues when merged, use [the special syntax for that](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) here.

### Main changes in the PR:

1. Added externalAppAuthentication.authenticateWithPowerPlatformConnectorPlugins
2. validateUrl util function to use in several capabilities

### Validation performed:

1. Added Unit tests
2. <Step 2>

### Unit Tests added:

> Unit tests are required for all changes. If no unit tests were added as part of this change, please explain why they aren't necessary.

<Yes/No>

### End-to-end tests added:

<Yes/No>

### Change file added:

> Ensure the change file meets the [formatting requirements](https://github.com/OfficeDev/microsoft-teams-library-js/blob/main/CONTRIBUTING.md#change-log-using-beachball).

<Yes/No>

### Related PRs:

> Remove this section if n/a

### Next/remaining steps:

> List the next or remaining steps in implementing the overall feature in subsequent PRs (or is the feature 100% complete after this?).

> Remove this section if n/a

- [ ] Item 1
- [ ] Item 2

### Screenshots:

> Remove this section if n/a

| Before     | After      |
| ---------- | ---------- |
| < image1 > | < image2 > |

### CODE REVIEW

File: packages/teams-js/src/internal/utils.ts [hunk 1]
```diff
@@ -416,6 +416,19 @@ export function validateId(id: string, errorToThrow?: Error): void {
   }
 }
 
+export function validateUrl(url: string, errorToThrow?: Error): void {
+  if (hasScriptTags(url)) { 
+    throw errorToThrow || new Error('Invaild Url');
```
Comment (erinha) [inline packages/teams-js/src/internal/utils.ts:421]:
```suggestion
    throw errorToThrow || new Error('Invalid Url');
```

File: packages/teams-js/src/internal/utils.ts [hunk 2]
```diff
@@ -416,6 +416,19 @@ export function validateId(id: string, errorToThrow?: Error): void {
   }
 }
 
+export function validateUrl(url: string, errorToThrow?: Error): void {
```
Comment (TrevorJoelHarris) [inline packages/teams-js/src/internal/utils.ts:419]: It seems odd to have this function and isValidHttpsUrl both exist in the utils file, will cause confusion for people deciding which one to use. Seems like there are a couple of options:

1) You could update `isValidHttpsURL` in this file to have this new logic that you're introducing here so everyone gets to use the new logic, or
2) You could make two separate functions but share the actual validation code between them both.

Would be good to not fork our Https URL validation code and have two different validation functions that validate differently.

File: change/@microsoft-teams-js-a11eae80-b237-4cf2-94ab-fb8e719e3dcb.json [hunk 3]
```diff
@@ -0,0 +1,7 @@
+{
+  "type": "minor",
+  "comment": "Added a new API `externalAppAuthentication.authenticateWithPowerPlatformConnectorPlugins`. It can be used to perform authentication with power platform connector plugins.",
```
Comment (AE-MS) [inline change/@microsoft-teams-js-a11eae80-b237-4cf2-94ab-fb8e719e3dcb.json:3]: >power platform [](http://example.com/codeflow?start=151&length=14)

Suggestion: "Power Platform"

File: packages/teams-js/src/private/externalAppAuthentication.ts [hunk 4]
```diff
@@ -498,6 +498,57 @@ export namespace externalAppAuthentication {
...
+   * @param titleId ID of the acquisition
+   * @param pluginId pluginId of the connector plugin
+   * @param signInUrl signInUrl for the connctor page listing the connector. This is optional
+   * @param oauthWindowParameters parameters for the signIn window
+   * @returns A promise that resolves when authentication succeeds and rejects with InvokeError on failure
+   */
+  export function authenticateWithPowerPlatformConnectorPlugins(
+    titleId: string,
+    pluginId: string,
+    signInUrl?: string,
```
Comment (AE-MS) [inline packages/teams-js/src/private/externalAppAuthentication.ts:516]: >string [](http://example.com/codeflow?start=16&length=6)

Can we accept this as a `URL` type instead of a `string` type?

**Review by AE-MS — CHANGES_REQUESTED** [review]
:clock1:

File: packages/teams-js/src/private/externalAppAuthentication.ts (hunk 4 above)
Comment (AE-MS) [reply packages/teams-js/src/private/externalAppAuthentication.ts:516]: We shouldn't be consistent with bad design just for the sake of consistency.

I'm not sure who "we" is regarding enforcing the rule but I can guarantee you that every PR I've ever done where I see URLs being passed as strings has received this comment. We should always use the most appropriate type to hold the data otherwise we might as well write directly in JavaScript.

I'll sign off on this because I don't know what other constraints under which you are operating and I don't want to hold up the whole effort for this. Feel free to discuss the tradeoffs further with @erinha if needed.

**Review by AE-MS — APPROVED** [review]
Signing off but feel free to discuss appropriate type choice further with @erinha as time permits.

File: packages/teams-js/src/private/externalAppAuthentication.ts (hunk 4 above)
Comment (erinha) [reply packages/teams-js/src/private/externalAppAuthentication.ts:516]: I agree that we should stop perpetuating bad habits. I am opening a work item to track modifying these private APIs to accept url parameters as `URL`s instead of strings.

@ashishsingh2105 do you know of any reason why you can't pass url as a URL type in the code that calls this API?

File: packages/teams-js/test/private/externalAppAuthentication.spec.ts [hunk 5]
```diff
@@ -738,6 +738,231 @@ describe('externalAppAuthentication', () => {
...
+          const invalidsingInUrl = 'https://pishingsite.com?<script>alert(1)</script>';
+          try {
+            await externalAppAuthentication.authenticateWithPowerPlatformConnectorPlugins(
+              titleId,
+              testPluginId,
+              invalidsingInUrl,
+              testPPCWindowParameters,
+            );
+          } catch (e) {
+            expect(e).toEqual(new Error('Invaild Url'));
```
Comment (erinha) [inline packages/teams-js/test/private/externalAppAuthentication.spec.ts:889]:
```suggestion
            expect(e).toEqual(new Error('Invalid Url'));
```

File: packages/teams-js/src/private/externalAppAuthentication.ts [hunk 6]
```diff
@@ -498,6 +498,59 @@ export namespace externalAppAuthentication {
...
+   * Limited to Microsoft-internal use
+   * @param titleId ID of the acquisition
+   * @param pluginId pluginId of the connector plugin
+   * @param signInUrl signInUrl for the connctor page listing the connector. This is optional
+   * @param oauthWindowParameters parameters for the signIn window
+   * @returns A promise that resolves when authentication succeeds and rejects with InvokeError on failure
+   */
+  export function authenticateWithPowerPlatformConnectorPlugins(
+    titleId: string,
+    pluginId?: string,
```
Comment (erinha) [inline packages/teams-js/src/private/externalAppAuthentication.ts:515]: Is this still necessary?

File: packages/teams-js/src/private/externalAppAuthentication.ts [hunk 7]
```diff
@@ -498,6 +498,59 @@ export namespace externalAppAuthentication {
...
+      validateId(pluginId, new Error('pluginId is Invalid.'));
+    }
+    if (signInUrl) {
+      validateUrl(signInUrl);
+    }
+
+    return sendMessageToParentAsync(
+      getApiVersionTag(
+        externalAppAuthenticationTelemetryVersionNumber,
+        ApiName.ExternalAppAuthentication_AuthenticateWithOauth2,
```
Comment (lakhveerk) [inline packages/teams-js/src/private/externalAppAuthentication.ts:536]: Api name needs an update to `authenticateWithPowerPlatformConnectorPlugins`, and also needed to add in ApiName in Telemetry.ts

File: packages/teams-js/src/private/externalAppAuthentication.ts [hunk 8]
```diff
@@ -498,6 +498,59 @@ export namespace externalAppAuthentication {
...
+    return sendMessageToParentAsync(
+      getApiVersionTag(
+        externalAppAuthenticationTelemetryVersionNumber,
+        ApiName.ExternalAppAuthentication_AuthenticateWithPowerPlatformConnectorPlugins,
+      ),
+      'externalAppAuthentication.authenticateWithPowerPlatformConnectorPlugins',
+      [
+        titleId,
+        pluginId,
+        signInUrl,
```
Comment (lakhveerk) [inline packages/teams-js/src/private/externalAppAuthentication.ts:542]: should this be changed back to string when sending over?

File: packages/teams-js/test/internal/utils.spec.ts [hunk 9]
```diff
@@ -188,6 +189,61 @@ describe('utils', () => {
...
+      try {
+        validateUrl(new URL(url));
+      } catch (error) {
+        expect(error).toEqual(new Error('Invalid Url'));
+      }
+    });
+    it('should throw invalid url error if it contains multiple script tag', async () => {
+      expect.assertions(1);
+      const url =
+        'https://example.com?id=1&param=script>alert("Hello, world!");</script>&val=3&param=<script>alert("Hello, world!");</script>';
```
Comment (shrshindeMSFT) [inline packages/teams-js/test/internal/utils.spec.ts:205]:
```suggestion
        'https://example.com?id=1&param=<script>alert("Hello, world!");</script>&val=3&param=<script>alert("Hello, world!");</script>';
```

File: packages/teams-js/src/internal/utils.ts [hunk 10]
```diff
@@ -416,9 +416,59 @@ export function validateId(id: string, errorToThrow?: Error): void {
...
-  return scriptRegex.test(id);
+  return scriptRegex.test(decodedInput);
+}
+
+/** 
+ * The decodeHTMLEntities function replaces HTML entities in the input string with their corresponding characters. 
+ * It uses a regular expression to match HTML entities and a replacement function to replace each match with the corresponding character from the entityMap object. 
+ * If a match is not found in the entityMap, it leaves the match as is.
+*/
+function decodeHTMLEntities(input: string): string {
```
Comment (shrshindeMSFT) [inline packages/teams-js/src/internal/utils.ts:459]: I understand the regex here but it is really hard to follow, a good alternative would be to:

``` 
function decodeHTMLEntities(input: string): string {
    // Define the entity map
    const entityMap: { [key: string]: string } = {
        '&lt;': '<',
        '&gt;': '>',
        '&amp;': '&',
        '&quot;': '"',
        '&#39;': "'",
        '&#x2F;': '/',
    };
    // Replace each entity with its corresponding value
    for (const entity in entityMap) {
        if (Object.prototype.hasOwnProperty.call(entityMap, entity)) {
            input = input.replace(new RegExp(entity, 'g'), entityMap[entity]);
        }
    } 
    return input;
}

As we won't require to keep track of the Regex here.

File: .vscode/settings.json [hunk 11]
```diff
@@ -12,5 +12,9 @@
   "prettier.singleQuote": true,
   "cSpell.words": ["adal", "uninitialize"],
   "jest.jestCommandLine": "pnpm test -- --",
-  "jest.autoRun": { "watch": "true", "onStartup": ["all-tests"] }
+  "jest.autoRun": {
+    "watch": "true",
+    "onStartup": ["all-tests"]
+  },
+  "CodeGPT.apiKey": "CodeGPT Plus Beta"
```
Comment (erinha) [inline .vscode/settings.json:19]: What caused this change?

File: packages/teams-js/src/internal/utils.ts [hunk 12]
```diff
@@ -416,9 +416,57 @@ export function validateId(id: string, errorToThrow?: Error): void {
...
+    ['&lt;', '<'],
+    ['&gt;', '>'],
+    ['&amp;', '&'],
+    ['&quot;', '"'],
+    ['&#39;', "'"],
+    ['&#x2F;', '/'],
+  ]);
+  entityMap.forEach((value, key) => {
+    input = input.replace(new RegExp(key, 'gi'), value);
+  });
```
Comment (erinha) [inline packages/teams-js/src/internal/utils.ts:468]: The latest version of the PR still appears to use RegExp... but the resolved comment suggests you planned to replace it with string.replace. Is that still your plan @ndangudubiyyam ?

File: packages/teams-js/test/internal/utils.spec.ts [hunk 13]
```diff
@@ -188,6 +189,97 @@ describe('utils', () => {
...
+    it('should throw invalid url error if it contains uppercase script tags', async () => {
+      expect.assertions(1);
+      const url = 'https://example.com?param=<script>alert("Hello, world!");</script>'.toLocaleUpperCase();
+      try {
+        validateUrl(new URL(url));
+      } catch (error) {
+        expect(error).toEqual(new Error('Invalid Url'));
+      }
+    });
+    it('should throw invalid url error if it contains camel case script tags', async () => {
```
Comment (erinha) [inline packages/teams-js/test/internal/utils.spec.ts:211]: nit: 
```suggestion
    it('should throw invalid url error if it contains mixed case script tags', async () => {
```

File: packages/teams-js/test/internal/utils.spec.ts [hunk 14]
```diff
@@ -188,6 +189,97 @@ describe('utils', () => {
...
+    it('should throw invalid url error if it contains HTML encoded script tags in upper case', async () => {
+      expect.assertions(1);
+      const url = 'https://example.com?param=&lt;script&gt;alert("Hello, world!");&lt;/script&gt;'.toLocaleUpperCase();
+      try {
+        validateUrl(new URL(url));
+      } catch (error) {
+        expect(error).toEqual(new Error('Invalid Url'));
+      }
+    });
+    it('should throw invalid url error if it contains HTML encoded script tags in camel case', async () => {
```
Comment (erinha) [inline packages/teams-js/test/internal/utils.spec.ts:248]: nit
```suggestion
    it('should throw invalid url error if it contains HTML encoded script tags in mixed case', async () => {
```

**Review by erinha — CHANGES_REQUESTED** [review]
Waiting on question about removal of RegExp usage in src/internal/utils.ts.

**Review by erinha — APPROVED** [review]
Looks good, thanks for addressing all my comments!

**Review by lakhveerk — APPROVED** [review]
